### ISSUE TITLE
Select App error

### ISSUE THREAD

**Issue by Daemonceltics** [dmayer/idb#91]
Error downloading file. 
/usr/local/lib/ruby/gems/2.3.0/gems/plist4r-1.2.2/lib/plist4r/plist.rb:294:in `open': No filename specified (RuntimeError)

**Comment by jimmyjammz99** [comment]
Yeah... I figured out why it is doing that... but it will be some work to fix it...

Modified the code a bit to insert some additional logging during the ssh_operations.rb.
under this section:
`    def download(remote_path, local_path = nil)
      begin
        if local_path.nil?
          **$log.info "Starting download #{@sftp} #{remote_path}"**
          @sftp.download! remote_path
        else
          **$log.info "Starting download #{@sftp} #{remote_path} to #{local_path}"**
          @sftp.download! remote_path, local_path
        end
      rescue
        $log.error "Error downloading file."
        return false
      end
      true
    end`

It will give you output like this in the log...
[INFO] 2017-06-22 15:42:09 :: idb ::  Starting download #<Net::SFTP::Session:0x00000001dccd90> /private/var/installd/Library/MobileInstallation/LastLaunchServicesMap.plist to /home/santoku/.idb/tmp//0B9AD0EC-2FBD-406D-8C5E-03F72173091F//private/var/installd/Library/MobileInstallation/LastLaunchServicesMap.plist
[ERROR] 2017-06-22 15:42:09 :: idb ::  Error downloading file.

the app.rb file defines this path inside this block of code:

`      if $device.ios_version >= 8
        if $device.ios_version == 8
          mapping_file = "/var/mobile/Library/MobileInstallation/LastLaunchServicesMap.plist"
        else
          mapping_file = "/private/var/installd/Library/MobileInstallation/LastLaunchServicesMap.plist"
        end
        local_mapping_file = cache_file mapping_file
        @services_map = IOS8LastLaunchServicesMapWrapper.new local_mapping_file
        @data_dir = @services_map.data_path_by_bundle_id @info_plist.bundle_identifier
        @keychain_access_groups = @services_map.keychain_access_groups_by_bundle_id @info_plist.bundle_identifier

      else
        @data_dir = @app_dir
      end`


The issue is iOS 10.x does not use the file **LastLaunchServicesMap.plist** anymore.  They are using a database for the list of installed applications. **/User/Library/FrontBoard/applicationState.db**

So basically a sql query will need to take place to grab the appropriate information idb is requesting...  Wonder if anyone has time to work through this...

Anyways... hope this points someone in the correct direction...

**Comment by dmayer** [comment]
thanks for getting this far. Will look into this.

**Comment by dmayer** [comment]
this should be fixed in the latest version. Are you still having issues?

**Comment by MuchiMuchiPink** [comment]
App listing works now, but idb crashes when I select an App.
 iOS 10.2
my os:  4.4.0-93-generic #116-Ubuntu 

`        from /var/lib/gems/2.3.0/gems/qtbindings-4.8.6.3/lib/Qt/qtruby4.rb:479:in `method_missing'
        from /var/lib/gems/2.3.0/gems/qtbindings-4.8.6.3/lib/Qt/qtruby4.rb:479:in `exec'
        from /var/lib/gems/2.3.0/gems/idb-2.10.1/lib/idb.rb:278:in `run'
        from /var/lib/gems/2.3.0/gems/idb-2.10.1/bin/idb:4:in `<top (required)>'
        from /usr/local/bin/idb:22:in `load'
        from /usr/local/bin/idb:22:in `<main>'
recv failed: Resource temporarily unavailable
`

**Comment by dmayer** [comment]
Are you able to share which app causes this? If its a free one I would install and test it.
Which ruby version are you using?
Also, is there more output you could paste? That does not seem to be the entire error.

**Comment by Daemonceltics** [comment]
Thanks a lot @dmayer

**Comment by MuchiMuchiPink** [comment]
DB Navigator, but it crashes on any App I tried.
ruby version is: ruby 2.3.4p301 (2017-03-30 revision 58214) [x86_64-linux-gnu]
But the bin points to an  older version (2.1.0). I might try to rebuild the gem first.

**Comment by ximerus** [comment]
I have same problem on app select. Thats my log

```
/Users/ximerus/.rvm/gems/ruby-2.4.1/gems/idb-2.10.3/lib/gui/app_details_group_box.rb:69:in `app_changed': undefined method `sub' for {"identifier"=>nil, "snapshots"=>nil, "$class"=>nil}:Hash (NoMethodError)
Did you mean?  sum
	from /Users/ximerus/.rvm/gems/ruby-2.4.1/gems/idb-2.10.3/lib/gui/app_tab_widget.rb:48:in `app_changed'
	from /Users/ximerus/.rvm/gems/ruby-2.4.1/gems/idb-2.10.3/lib/gui/main_tab_widget.rb:163:in `app_changed'
	from /Users/ximerus/.rvm/gems/ruby-2.4.1/gems/idb-2.10.3/lib/idb.rb:122:in `block in init_ui'
	from /Users/ximerus/.rvm/gems/ruby-2.4.1/gems/qtbindings-0.0.0.0/lib/Qt/qtruby4.rb:2470:in `invoke'
	from /Users/ximerus/.rvm/gems/ruby-2.4.1/gems/idb-2.10.3/lib/gui/global_app_details_group_box.rb:61:in `qt_metacall'
	from /Users/ximerus/.rvm/gems/ruby-2.4.1/gems/idb-2.10.3/lib/gui/global_app_details_group_box.rb:61:in `app_changed'
	from /Users/ximerus/.rvm/gems/ruby-2.4.1/gems/idb-2.10.3/lib/gui/global_app_details_group_box.rb:61:in `block (2 levels) in initialize'
	from /Users/ximerus/.rvm/gems/ruby-2.4.1/gems/qtbindings-0.0.0.0/lib/Qt/qtruby4.rb:2470:in `invoke'
	from /Users/ximerus/.rvm/gems/ruby-2.4.1/gems/idb-2.10.3/lib/gui/app_list_dialog.rb:24:in `qt_metacall'
	from /Users/ximerus/.rvm/gems/ruby-2.4.1/gems/idb-2.10.3/lib/gui/app_list_dialog.rb:24:in `method_missing'
	from /Users/ximerus/.rvm/gems/ruby-2.4.1/gems/idb-2.10.3/lib/gui/app_list_dialog.rb:24:in `block in initialize'
	from /Users/ximerus/.rvm/gems/ruby-2.4.1/gems/qtbindings-0.0.0.0/lib/Qt/qtruby4.rb:2470:in `invoke'
	from /Users/ximerus/.rvm/gems/ruby-2.4.1/gems/qtbindings-0.0.0.0/lib/Qt/qtruby4.rb:917:in `qt_metacall'
	from /Users/ximerus/.rvm/gems/ruby-2.4.1/gems/qtbindings-0.0.0.0/lib/Qt/qtruby4.rb:917:in `method_missing'
	from /Users/ximerus/.rvm/gems/ruby-2.4.1/gems/qtbindings-0.0.0.0/lib/Qt/qtruby4.rb:917:in `exec'
	from /Users/ximerus/.rvm/gems/ruby-2.4.1/gems/idb-2.10.3/lib/gui/global_app_details_group_box.rb:63:in `block in initialize'
	from /Users/ximerus/.rvm/gems/ruby-2.4.1/gems/qtbindings-0.0.0.0/lib/Qt/qtruby4.rb:2470:in `invoke'
	from /Users/ximerus/.rvm/gems/ruby-2.4.1/gems/qtbindings-0.0.0.0/lib/Qt/qtruby4.rb:479:in `qt_metacall'
	from /Users/ximerus/.rvm/gems/ruby-2.4.1/gems/qtbindings-0.0.0.0/lib/Qt/qtruby4.rb:479:in `method_missing'
	from /Users/ximerus/.rvm/gems/ruby-2.4.1/gems/qtbindings-0.0.0.0/lib/Qt/qtruby4.rb:479:in `exec'
	from /Users/ximerus/.rvm/gems/ruby-2.4.1/gems/idb-2.10.3/lib/idb.rb:278:in `run'
	from /Users/ximerus/.rvm/gems/ruby-2.4.1/gems/idb-2.10.3/bin/idb:4:in `<top (required)>'
	from /Users/ximerus/.rvm/gems/ruby-2.4.1/bin/idb:23:in `load'
	from /Users/ximerus/.rvm/gems/ruby-2.4.1/bin/idb:23:in `<main>'
	from /Users/ximerus/.rvm/gems/ruby-2.4.1/bin/ruby_executable_hooks:15:in `eval'
	from /Users/ximerus/.rvm/gems/ruby-2.4.1/bin/ruby_executable_hooks:15:in `<main>'
recv failed: Resource temporarily unavailable
```

**Comment by whoot** [comment]
Same problem here!
When selecting an app, idb crashes. This is my log:

```
[DEBUG] 2017-11-13 17:17:59 :: idb ::  Executing blocking SSH command: /usr/bin/ldid -e '/private/var/containers/Bundle/Application/6B9C9D66-A9C4-4CB7-B9E0-2ACBF17040B7//DamnVulnerableIOSApp.app/DamnVulnerableIOSApp'
/Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/plist4r-1.2.2/lib/plist4r/plist.rb:82:in `from_string': Unknown plist format for string:  (RuntimeError)
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/plist4r-1.2.2/lib/plist4r/plist.rb:280:in `block in parse_opts'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/plist4r-1.2.2/lib/plist4r/plist.rb:277:in `each'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/plist4r-1.2.2/lib/plist4r/plist.rb:277:in `parse_opts'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/plist4r-1.2.2/lib/plist4r/plist.rb:53:in `initialize'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/plist4r-1.2.2/lib/plist4r.rb:20:in `new'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/plist4r-1.2.2/lib/plist4r.rb:20:in `new'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/idb-2.10.3/lib/lib/ios10_application_state_db_wrapper.rb:28:in `entitlements_by_binary'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/idb-2.10.3/lib/lib/app.rb:300:in `entitlements'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/idb-2.10.3/lib/gui/default_protection_class_group_widget.rb:23:in `update'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/idb-2.10.3/lib/gui/plist_file_widget.rb:52:in `setup'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/idb-2.10.3/lib/gui/local_storage_tab_widget.rb:25:in `setup'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/idb-2.10.3/lib/gui/main_tab_widget.rb:95:in `enableLocalStorage'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/idb-2.10.3/lib/gui/main_tab_widget.rb:159:in `app_changed'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/idb-2.10.3/lib/idb.rb:122:in `block in init_ui'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/qtbindings-4.8.6.4/lib/Qt/qtruby4.rb:2470:in `invoke'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/idb-2.10.3/lib/gui/global_app_details_group_box.rb:61:in `qt_metacall'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/idb-2.10.3/lib/gui/global_app_details_group_box.rb:61:in `app_changed'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/idb-2.10.3/lib/gui/global_app_details_group_box.rb:61:in `block (2 levels) in initialize'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/qtbindings-4.8.6.4/lib/Qt/qtruby4.rb:2470:in `invoke'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/idb-2.10.3/lib/gui/app_list_dialog.rb:24:in `qt_metacall'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/idb-2.10.3/lib/gui/app_list_dialog.rb:24:in `method_missing'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/idb-2.10.3/lib/gui/app_list_dialog.rb:24:in `block in initialize'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/qtbindings-4.8.6.4/lib/Qt/qtruby4.rb:2470:in `invoke'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/qtbindings-4.8.6.4/lib/Qt/qtruby4.rb:917:in `qt_metacall'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/qtbindings-4.8.6.4/lib/Qt/qtruby4.rb:917:in `method_missing'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/qtbindings-4.8.6.4/lib/Qt/qtruby4.rb:917:in `exec'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/idb-2.10.3/lib/gui/global_app_details_group_box.rb:63:in `block in initialize'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/qtbindings-4.8.6.4/lib/Qt/qtruby4.rb:2470:in `invoke'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/qtbindings-4.8.6.4/lib/Qt/qtruby4.rb:479:in `qt_metacall'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/qtbindings-4.8.6.4/lib/Qt/qtruby4.rb:479:in `method_missing'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/qtbindings-4.8.6.4/lib/Qt/qtruby4.rb:479:in `exec'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/idb-2.10.3/lib/idb.rb:278:in `run'
	from /Users/admin/.rvm/rubies/ruby-2.4.1/lib/ruby/gems/2.4.0/gems/idb-2.10.3/bin/idb:4:in `<top (required)>'
	from /Users/admin/.rvm/gems/ruby-2.4.1/bin/idb:23:in `load'
	from /Users/admin/.rvm/gems/ruby-2.4.1/bin/idb:23:in `<main>'
	from /Users/admin/.rvm/gems/ruby-2.4.1/bin/ruby_executable_hooks:15:in `eval'
	from /Users/admin/.rvm/gems/ruby-2.4.1/bin/ruby_executable_hooks:15:in `<main>'
recv failed: Resource temporarily unavailable
``

**Comment by whoot** [comment]
I found a **workaround** for the app crashing issue.
It seems like there is something wrong with the iOS 10 support.
Revert this and all subsequent commits and idb will work again!

You have to do the following steps:
```
git clone https://github.com/dmayer/idb
cd idb/
git revert 038355497091b24c53596817b8818d2b2bc18e4b
git revert 799a0e6565c17f72dce5a7e9f1a5738abbd08ae2
git revert 092929aa51b941e003f7e4d162b550171da8d66a
git revert c8dcb9f091fc7f706af60b40c5a041299c3dcf35
git revert c8d025012fa7eacaae688a53e22078ccf4cebe15

bundle install
bundle exec idb
```

**WARNING: This will drop the support for iOS 10!**

**Comment by MuchiMuchiPink** [comment]
Thanks for the workaround, however I need idb working on iOS 10. ;)

**Comment by jtof-fap** [comment]
After some researchs, I found the root cause of crash in iOS 10. I use an `iPhone 5c 10.3.3 (32bit/Helix)` and the application crashs when I select an application in list because IDB failed to get the correct `data_dir`  from `/User/Library/FrontBoard/applicationState.db`.

The error comes from the value of `kvs.key` on the sql request in `/lib/lib/ios10_application_state_db_wrapper.rb` :

    stmnt = db.prepare "SELECT kvs.value FROM application_identifier_tab left join kvs on application_identifier_tab.id = kvs.application_identifier where kvs.key = 2 and application_identifier_tab.application_identifier='#{bundle_id}'"

On my iPhone, the `app_dir` and `data_dir` are in the blob associated to the kvs.key number `1`. Just replace `2`  (_XBApplicationSnapshotManifest_) by `1` (_compatibilityInfo_) in this request work like a charm for me on iOS 10.3.3. 

You could test the following command to find an application `data_dir` (ex: facebook)

    # sqlite3 /tmp/applicationState.db "select quote(value) from kvs inner join application_identifier_tab on kvs.application_identifier = application_identifier_tab.id where application_identifier_tab.application_identifier like '%facebook%' and kvs.key = 1;" | cut -d "'" -f2 | xxd -r -p | strings | sed -n 's/T\(\/private\/var\/mobile\/Containers\/Data.*\)_/\1/p'
    /private/var/mobile/Containers/Data/Application/B3D96422-D800-4040-9047-A8BC8E977929

The solution is to test the others `kvs.key` values if the `data_dir` is null.

I hope it will help.

**Comment by mattymcfatty** [comment]
oh, I think they changed the directories of where apps are stored in iOS > 10. Could this be the issue?
Applications have been moved in iOS 10. They're now in
/private/var/containers/Bundle/Application/

**Comment by jtof-fap** [comment]
No, it is not the issue.

When you select an application in the list, IDB need to get 2 directories :
 * The app bundle directory : `in /private/var/containers/Bundle/Application/{UUID}`
 * The data directory which contain documents, cache, preferences, etc. in `/private/var/mobile/Containers/Data/Application/{XXX different of UUID}`.

As mentionned previously, before IOS 10 all these information was in `LastLaunchServicesMap.plist` file. Since IOS 10, these information are stored in binary format in  a sqllite database : `/User/Library/FrontBoard/applicationState.db`.

IDB is compatible with IOS 10 and get correctly `app_dir`and `data_dir` in `applicationState.db` and probabely works with some IOS 10 versions.

In my case, the SQL request in the code return nothing for the `data_dir` and, in trying to update the IHM label `Data directory` with null value, the application crashs with the following error message : 

    .../idb-2.10.3/lib/gui/app_details_group_box.rb:69:in `app_changed': undefined method `sub' for {"identifier"=>nil, "snapshots"=>nil, "$class"=>nil}:Hash (NoMethodError)
    Did you mean?  sum
    ....

By adapting the request, in my case (or by coding in hard the correct data_dir value), IDB don't crash and  perfectly works with my phone in IOS 10.

I don't know if it is due to the architecture (32bit vs 64bit) of my phone `5C` or my IOS version `10.3.3` but in my case, the `data_dir` value is in kvs.key `1` (compatibilityInfo)  and not in kvs.key `2` (XBApplicationSnapshotManifest) in the `applicationState.db` database. Maybe that the kvs.key values in database vary between phones and by using name (compatibilityInfo) instead of kvs.key values it would fix the problem for everybody. I don't have other phones in order to test.

**Comment by mattymcfatty** [comment]
Ah. So helpful! Thanks @jtof-fap !

I grepped "data_dir" to see if I could hard-code it like you did, but I got a few diff files. Did you hard code the "data_dir" value in "app_details_group_box.rb" or a different file? It would be clutch if I got this working for my current test. 

Also, I'm on a 5s with iOS version 10.1.1 and IDB is on my 64-bit Linux host. Is my rig different enough from yours to help fix the problem? I'd love to collaborate to see if we can fix this for everyone.

**Comment by mattymcfatty** [comment]
oh nvm. I guess I should read higher up.... I've got the filename. Let me know if my version(s) would be any help nailing this down.

**Comment by jtof-fap** [comment]
Had you try to set : `kvs.key = 1` (instead of 2) on the SQL request in `/lib/lib/ios10_application_state_db_wrapper.rb` : line 44 ?

Except for testing, hardcoding `data_dir` isn't good solution but it works. If you want hardcode the value, the best place is in `/lib/lib/app.rb` but you will break the flexibility of app choice for future.

**Comment by mattymcfatty** [comment]
Awesome! Using your fix in /lib/lib/ios10_application_state_db_wrapper.rb I got it working (changed the 2 to a 1). 

Maybe i'll dust off the GIT manual and submit a fix that checks if results are empty, if so, try switching 2 to 1. Gotta finish this iOS test first tho...

Thanks for the help!

**Comment by jtof-fap** [comment]
I don't submit a fix because I think kvs.key=2 (set by default) seems to be work on certain phones and setting permanently to 1 will probably break the compatibility with these phones...

Maybe that the kvs.key values in database vary between phones and by using name (compatibilityInfo) instead of kvs.key values it would fix the problem for everybody. I don't have other phones in order to test.

**Comment by bensh** [comment]
Changing the kvs.key to 1 didn't work for me.

ruby 2.4.2
idb 2.10.3
iPad Air - iPad4,1
ios 10.0.2

**Comment by jtof-fap** [comment]
Thanks for your return. Try the fork of zk-ncc : [https://github.com/dmayer/idb/pull/105/commits/038413e444eccbe60455767f8e12e65acb19b040](https://github.com/dmayer/idb/pull/105/commits/038413e444eccbe60455767f8e12e65acb19b040)

**Comment by bensh** [comment]
Tried the fork with the code change, still crashing with recv failed: resource temporarily unavailable

Tried with Lloyds Bank and EdgeVis Mobile, both free apps on the store.

**Comment by whoot** [comment]
I'm honestly not sure if IDB will be supported for much longer, since the response time and the support in general is awful..
Meanwhile I simply switched to another tool (*cough* Needle *cough*), which works very well and provides the same features (and more).

**Comment by MuchiMuchiPink** [comment]
I've got Needle and Objection.

**Comment by bensh** [comment]
Shame this thread seems to have died with no update. idb was a very good tool for quick sanity checking with easily read output

**Comment by MuchiMuchiPink** [comment]
@bensh maybe look into passionfruit. https://github.com/chaitin/passionfruit
